File: python/tree/PQueue.py
# 最大优先队列
class PriorityQueue_max:
    def __init__(self):
        self.array = [0] * 32
        self.size = 0

    # 入队
    def enQueue(self, key):
        if self.size >= len(self.array):
            self.resize()
        self.array[self.size] = key
        self.size += 1
        self.upAdjust_max()

    # 出队
    def deQueue(self):
        if self.size <= 0:
            print("the queue is empty!")
            return None
        # 获取堆顶元素
        head = self.array[0]
        # 将最后一个元素移动到堆顶
        self.array[0] = self.array[self.size - 1]
        self.size -= 1
        self.downAdjust_max()
        return head

    def resize(self):
        # 队列容量翻倍
        new_array = [0] * (len(self.array) * 2)
        for i in range(self.size):
            new_array[i] = self.array[i]
        self.array = new_array
    
    def upAdjust_max(self):
        childIndex = self.size - 1
        parentIndex = (childIndex - 1) //2
        #temp保存插入的叶子节点值，用于最后赋值
        temp = self.array[childIndex] 
        while childIndex > 0 and temp > self.array[parentIndex]:
            #无须真正交换，单向赋值即可
            self.array[childIndex] = self.array[parentIndex]
            childIndex = parentIndex
            parentIndex = (parentIndex - 1) //2
        self.array[childIndex] = temp
    
    def downAdjust_max(self):
        parentIndex = 0
        #temp保存父节点的值，用于最后赋值
        temp = self.array[parentIndex]
        childIndex = 1
        while childIndex < self.size:
            # 如果有右孩子，且右孩子的值大于左孩子的值，则定位到右孩子
            if childIndex + 1 < self.size and self.array[childIndex + 1] > self.array[childIndex]:
                childIndex += 1
            # 如果父节点大于任何一个孩子节点，则直接跳出
            if temp >= self.array[childIndex]:
                break
            # 无须真正交换，单向赋值即可
            self.array[parentIndex] = self.array[childIndex]
            parentIndex = childIndex
            childIndex = 2 * childIndex + 1
        self.array[parentIndex] = temp



# 最小优先队列
class PriorityQueue_min:
    def __init__(self):
        self.array= [0] *32
        self.size = 0

    # 入队
    def enQueue(self, key):
        if self.size >= len(self.array):
            self.resize()
        self.array[self.size] = key
        self.size += 1
        self.upAdjust_min()

    # 出队
    def deQueue(self):
        if self.size <= 0:
            print("the queue is empty!")
            return None
        # 获取堆顶元素
        head = self.array[0]
        # 将最后一个元素移动到堆顶
        self.array[0] = self.array[self.size-1]
        self.size -= 1
        #self.array.pop()
        self.downAdjust_min()
        return head

    def resize(self):
        # 队列容量翻倍
        new_array = [0] * (len(self.array) * 2)
        for i in range(self.size):
            new_array[i] = self.array[i]
        self.array = new_array
    
    def upAdjust_min(self):
        childIndex = self.size -1
        parentIndex = (childIndex -1) //2
        #temp保存插入的叶子节点值，用于最后赋值
        temp = self.array[childIndex] 
        while childIndex >0 and temp < self.array[parentIndex]:
            self.array[childIndex] = self.array[parentIndex]
            childIndex = parentIndex
            parentIndex = (parentIndex -1) //2
        self.array[childIndex] =temp
    
    def downAdjust_min(self):
        parentIndex = 0
        childIndex =1
        #temp保存父节点的值，用于最后赋值
        temp = self.array[parentIndex]
        while childIndex <self.size:
            #如果有右孩子，且右孩子的值小于左孩子，则定位到右孩子
            if childIndex +1< self.size and self.array[childIndex+1]<self.array[childIndex]:
                childIndex +=1
            #如果父节点小于所有孩子节点，直接跳出
            if temp <self.array[childIndex]:
                break
            #无须真正交换，单向赋值即可
            self.array[parentIndex] = self.array[childIndex]
            parentIndex = childIndex
            childIndex = 2 * childIndex +1
        self.array[parentIndex] =temp

if __name__ == "__main__":
    priorityQueue1 = PriorityQueue_max()
    priorityQueue1.enQueue(3)
    priorityQueue1.enQueue(5)
    priorityQueue1.enQueue(10)
    priorityQueue1.enQueue(2)
    priorityQueue1.enQueue(7)
    print("出队元素1：")
    print(priorityQueue1.deQueue())
    print("出队元素2：")
    print(priorityQueue1.deQueue())

    priorityQueue2 = PriorityQueue_min()
    priorityQueue2.enQueue(3)
    priorityQueue2.enQueue(5)
    priorityQueue2.enQueue(10)
    priorityQueue2.enQueue(2)
    priorityQueue2.enQueue(7)
    print("出队元素3：")
    print(priorityQueue2.deQueue())
    print("出队元素4：")
    print(priorityQueue2.deQueue())

File: C++/tree/PQueue.cpp
#include<iostream>
using namespace std;

class PriorityQueue_max {
public:
    PriorityQueue_max() {
        array = new int[5];
        size = 0;
        length = 5;
    }

    ~PriorityQueue_max() {
        delete[] array;
    }

    void enQueue(int key) {
        if (size >= length) {
            resize();
        }
        array[size++] = key;
        upAdjust();
    }

    int deQueue() {
        if (size <= 0) {
            cout << "The queue is empty!" << endl;
            return -1;
        }
        int head = array[0];
        array[0] = array[--size];
        downAdjust();
        return head;
    }

private:
    int *array;
    int size;
    int length;

    void upAdjust() {
        int childIndex = size - 1;
        int parentIndex = (childIndex - 1) / 2;
        int temp = array[childIndex];
        while (childIndex > 0 && temp > array[parentIndex]) {
            array[childIndex] = array[parentIndex];
            childIndex = parentIndex;
            parentIndex = (parentIndex - 1) / 2;
        }
        array[childIndex] = temp;
    }

    void downAdjust() {
        int parentIndex = 0;
        int childIndex = 1;
        int temp = array[parentIndex];
        while (childIndex < size) {
            if (childIndex + 1 < size && array[childIndex + 1] > array[childIndex]) {
                childIndex++;
            }
            if (temp > array[childIndex]) {
                break;
            }
            array[parentIndex] = array[childIndex];
            parentIndex = childIndex;
            childIndex = 2 * childIndex + 1;
        }
        array[parentIndex] = temp;
    }

    void resize() {
        int newSize = length * 2;
        int *newArray = new int[newSize];
        for (int i = 0; i < size; i++) {
            newArray[i] = array[i];
        }
        delete[] array;
        array = newArray;
        length = newSize;
    }
};

class PriorityQueue_min {
public:
    PriorityQueue_min() {
        array = new int[5];
        size = 0;
        length = 5;
    }

    ~PriorityQueue_min() {
        delete[] array;
    }

    void enQueue(int key) {
        if (size >= length) {
            resize();
        }
        array[size++] = key;
        upAdjust();
    }

    int deQueue() {
        if (size <= 0) {
            cout << "The queue is empty!" << endl;
            return -1;
        }
        int head = array[0];
        array[0] = array[--size];
        downAdjust();
        return head;
    }

private:
    int *array;
    int size;
    int length;

    void upAdjust() {
        int childIndex = size - 1;
        int parentIndex = (childIndex - 1) / 2;
        int temp = array[childIndex];
        while (childIndex > 0 && temp < array[parentIndex]) {
            array[childIndex] = array[parentIndex];
            childIndex = parentIndex;
            parentIndex = (parentIndex - 1) / 2;
        }
        array[childIndex] = temp;
    }

    void downAdjust() {
        int parentIndex = 0;
        int childIndex = 1;
        int temp = array[parentIndex];
        while (childIndex < size) {
            if (childIndex + 1 < size && array[childIndex + 1] < array[childIndex]) {
                childIndex++;
            }
            if (temp < array[childIndex]) {
                break;
            }
            array[parentIndex] = array[childIndex];
            parentIndex = childIndex;
            childIndex = 2 * childIndex + 1;
        }
        array[parentIndex] = temp;
    }

    void resize() {
        int newSize = length * 2;
        int *newArray = new int[newSize];
        for (int i = 0; i < size; i++) {
            newArray[i] = array[i];
        }
        delete[] array;
        array = newArray;
        length = newSize;
    }
};

int main() {
    PriorityQueue_max pqueue_max;
    pqueue_max.enQueue(3);
    pqueue_max.enQueue(5);
    pqueue_max.enQueue(10);
    pqueue_max.enQueue(2);
    pqueue_max.enQueue(7);
    cout << "最大优先队列的出队元素：" << pqueue_max.deQueue() << endl;
    cout << "最大优先队列的出队元素：" << pqueue_max.deQueue() << endl;

    PriorityQueue_min pqueue_min;
    pqueue_min.enQueue(3);
    pqueue_min.enQueue(5);
    pqueue_min.enQueue(10);
    pqueue_min.enQueue(2);
    pqueue_min.enQueue(7);
    cout << "最小优先队列的出队元素：" << pqueue_min.deQueue() << endl;
    cout << "最小优先队列的出队元素：" << pqueue_min.deQueue() << endl;

    return 0;
}
